<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Lesson_2(52)</title>
</head>
<style>
    .description {
        width: 600px;
    }

    .view {
        width: 300px;
    }

    div {
        width: 200px;
    }

    table, th, td {
        padding: 10px;
        border: 1px solid;
        border-collapse: collapse;
    }
</style>
<body>
<h1>Выбрасывание исключительных ситуаций в JavaScript</h1>
<table border="1">
    <tr>
        <td class="description">
            <p>
                В JavaScript попытка извлечь корень из отрицательного числа не приводит к выбрасыванию исключения:
            </p>
            <p>
                let result = Math.sqrt(-1);
                console.log(result); // выведет NaN
            </p>
            <p>
                Напишите свою функцию, которая будет извлекать корень из числа и при этом выбрасывать исключение,
                если корень извлекается из отрицательного числа.
            </p>
        </td>
        <td colspan="2" class="view" id="task_1">
            <input type="number">
            <button id="but_1">извлечь</button>
        </td>
    </tr>
</table>
<h1>Типы исключений</h1>
<table border="1">
    <tr>
        <td class="description">
            <p>
                Выбросите исключение с типом TypeError.
            </p>
        </td>
        <td colspan="2" class="view" id="task_2">
            <button id="but_2">Выбросить исключение</button>
        </td>
    </tr>
    <tr>
        <td class="description">
            <p>
                Выбросите исключение с типом SyntaxError и RangeError.
                Поймайте эти исключения с помощью одного блока try.
                В блоке catch выведите разные сообщения об ошибке для исключений разных типов.
            </p>
        </td>
        <td colspan="2" class="view" id="task_3">
            <button id="but_3_1">Выбросить исключение SyntaxError</button>
            <button id="but_3_2">Выбросить исключение RangeError</button>
        </td>
    </tr>
</table>
<h1>Свои типы исключений</h1>
<table border="1">
    <tr>
        <td class="description">
            <p>
                Выше я мы сделали функцию, выбрасывающую исключение при делении на ноль.
                Переделайте эту функцию так, чтобы она выбрасывала исключение с каким-нибудь придуманными
                нами типом, например, DivisionByZeroError.
            </p>
        </td>
        <td colspan="2" class="view" id="task_4">
            <input type="number" id="one">
            <input type="number" id="two">
            <button id="but_4">Разделить</button>
        </td>
    </tr>
    <tr>
        <td class="description">
            <p>
                Выше вы делали функцию, выбрасывающую исключение при попытке извлечь корень из отрицательного числа.
                Переделайте вашу функцию так, чтобы она выбрасывала исключение с придуманным вами типом. Хорошо
                подумайте над названием исключения, чтобы это название было удачным.
            </p>
        </td>
        <td colspan="2" class="view" id="task_5">
            <input type="number" id="three">
            <button id="but_5">Вычислить квадратный корень</button>
        </td>
    </tr>
</table>
<script>
    "use strict";
    function mySqrt(number) {
        if (number < 0) {
            throw new Error('Нельзя извлекать квадратный корень из отрицательного числа');
        } else {
            return Math.sqrt(number);
        }
    }

    document.querySelector('#but_1').addEventListener('click', () => {
        let n = document.querySelector('#task_1 input').value;
        let p = document.createElement('p');
        p.textContent = mySqrt(n);
        document.querySelector('#task_1').appendChild(p);
    });

    document.querySelector('#but_2').addEventListener('click', () => {
        try {
            throw new TypeError('Исключение!');
        } catch (error) {
            console.log(error.name);
            console.log(error.message);
        }
    });

    function testException(arr) {
        try {
            if (!Array.isArray(arr)) {
                throw new SyntaxError('Был передан не массив');
            }

            for (let i = 0; i < 5; i++) {
                let n = arr[i];
                if (n == undefined) {
                    throw new RangeError('Недостаточно элементов в массиве');
                }
            }
        } catch (error) {
            if (error.name == 'SyntaxError') {
                console.log('SyntaxError');
                console.log(error.message);
            } else if (error.name == 'RangeError') {
                console.log('RangeError');
                console.log(error.message);
            } else {
                console.log(error.name);
                console.log(error.message);
            }
        }
    }

    document.querySelector('#but_3_1').addEventListener('click', () => {
        testException({1: 1, 2: 2, 3: 3});
    });
    document.querySelector('#but_3_2').addEventListener('click', () => {
        let arr = [1, 2, 3];
        testException(arr);
    });

    function div(a, b) {
        if (b != 0) {
            return a / b;
        } else {
            throw {name: 'DivisionByZeroError', message: 'делить на ноль нельзя'};
        }
    }
    document.querySelector('#but_4').addEventListener('click', () => {
        let p = document.createElement('p');
        let one = document.querySelector('#one');
        let two = document.querySelector('#two');
        try {
            p.textContent = div(one.value, two.value);
            document.querySelector('#task_4').appendChild(p);

        } catch (error) {
            p.textContent = error.name + ' ' + error.message;
            document.querySelector('#task_4').appendChild(p);
        }
    });

    function mySqrtNew(number) {
        if (number < 0) {
            throw {name: 'ImpossibleException', message:'Нельзя извлекать квадратный корень из отрицательного числа'};
        } else {
            return Math.sqrt(number);
        }
    }
    document.querySelector('#but_5').addEventListener('click', () => {
        let p = document.createElement('p');
        let number = document.querySelector('#three');
        try {
            p.textContent = mySqrtNew(number.value);
            document.querySelector('#task_5').appendChild(p);
        } catch (error) {
            p.textContent = error.name + ' ' + error.message;
            document.querySelector('#task_5').appendChild(p);
        }
    });
</script>
</body>
</html>